lib/msprime.c

```diff
@@ -41,6 +42,51 @@
 #define MSP_STATE_SIMULATING 2
 #define MSP_STATE_DEBUGGING 3
 
+/* Draw a random variable from a truncated Beta(a, b) distribution,
+ * by rejecting draws above the truncation point x.
+ */
+static double
+ran_inc_beta_rej(gsl_rng *r, double a, double b, double x)
+{
+    double ret;
+    do {
+        ret = gsl_ran_beta(r, a, b);
+    } while(ret > x);
+    return ret;
+}
+
+/* Draw a random variable from a truncated Beta(a, b) distribution,
+ * using the inverse transform sampling method. A uniform variable
+ * is drawn and then transformed into a Beta(a, b) distribution using
+ * the inverse CDF. Truncation is obtained by defining the upper bound
+ * of the uniform variable using the incomplete beta function. I.e.
+ *      upper_bound = gsl_sf_beta_inc(a, b, x),
+ * where x is the truncation point.
+ */
+static double
+ran_inc_beta_its(gsl_rng *r, double a, double b, double upper_bound)
+{
+    double u = gsl_ran_flat(r, 0, upper_bound);
+    return gsl_cdf_beta_Pinv(u, a, b);
```

Comment:
Is it obvious that this is the right thing to do? A naive version of the inverse CDF method would use a `u ~ U(0,1)` uniform variate, and the CDF

`F(x) = B(x; a, b) / B(T; a, b)`,

where `B(x; a, b)` is the incomplete Beta function with parameters `a` and `b`, and `T` is the truncation point. Setting `F(x) = u` and doing a little algebra gives

`B(x; a, b) / B(1; a, b) = u B(T; a, b) / B(1; a, b)`,

which suggests that the right way to sample is 

`double u = gsl_rng_uniform(gen) * gsl_sf_beta_inc(alpha, beta, T) / gsl_sf_beta(alpha, beta);
`return gsl_cdf_beta_Pinv(u , alpha, beta);`

Is your way equivalent? If it is, it is definitely simpler.

Reply:
I followed the approach by Wikipedia et al. https://en.wikipedia.org/wiki/Inverse_transform_sampling#Truncated_distribution

I generated 10000 samples for each of `ran_inc_beta_rej` and `ran_inc_beta_its`, sorted the samples, and found their pearson correlation was >0.999 for a few different values of alpha and beta. See the ran_inc_beta.c.txt attached in an earlier comment.

Reply:
I think gsl_sf_beta_inc is already normalised. In which case your suggestion and my code are equivalent.
https://www.gnu.org/software/gsl/doc/html/specfunc.html#c.gsl_sf_beta_inc

Reply:
Ah, you're right. That's what I was missing, and makes your way obviously better. Thanks!

Reply:
I feel like framing this discussion as an example of "How Scientific Software Development Should Be Done"!